include/Surelog/Common/PathId.h

```diff
@@ -0,0 +1,140 @@
+#ifndef SURELOG_PATHID_H
+#define SURELOG_PATHID_H
+#pragma once
+
+#include <Surelog/Common/SymbolId.h>
+
+#include <cstdint>
+#include <filesystem>
+#include <istream>
+#include <ostream>
+#include <set>
+#include <string_view>
+#include <unordered_set>
+#include <vector>
+
+#if !defined(PATHID_DEBUG_ENABLED)
+#if defined(DEBUG) || defined(_DEBUG)
+#define PATHID_DEBUG_ENABLED 1
+#else
+#define PATHID_DEBUG_ENABLED 0
+#endif
+#endif
+
+namespace SURELOG {
+typedef uint32_t RawPathId;
+inline static constexpr RawPathId BadRawPathId = 0;
+inline static constexpr std::string_view BadRawPath = BadRawSymbol;
+
+class SymbolTable;
+
+class PathId final {
+ public:
+#if PATHID_DEBUG_ENABLED
+  PathId()
+      : m_symbolTable(nullptr), m_id(BadRawPathId), m_value(BadRawPath) {}
+  PathId(const SymbolTable *const symbolTable, RawPathId id,
+         std::string_view value)
+      : m_symbolTable(symbolTable), m_id(id), m_value(value) {}
+  PathId(const PathId &rhs)
+      : PathId(rhs.m_symbolTable, rhs.m_id, rhs.m_value) {}
+  PathId(const SymbolTable *const symbolTable, SymbolId id)
+      : PathId(symbolTable, id.id, id.value) {}
+#else
+  PathId() : m_symbolTable(nullptr), m_id(BadRawPathId) {}
+  PathId(const SymbolTable *const symbolTable, RawPathId id,
+         std::string_view value)
+      : m_symbolTable(symbolTable), m_id(id) {}
+  PathId(const PathId &rhs)
+      : PathId(rhs.m_symbolTable, rhs.m_id, BadRawPath) {}
+  PathId(const SymbolTable *const symbolTable, SymbolId id)
+      : PathId(symbolTable, id.id, BadRawPath) {}
+#endif
+
+  PathId &operator=(const PathId &rhs) {
+    if (this != &rhs) {
+      m_symbolTable = rhs.m_symbolTable;
+      m_id = rhs.m_id;
+#if PATHID_DEBUG_ENABLED
+      m_value = rhs.m_value;
+#endif
+    }
+    return *this;
+  }
+
+  const SymbolTable *getSymbolTable() const { return m_symbolTable; }
```

Comment:
It is unclear why a path-ID needs to know about the SymbolTable. It sounds like a leaky abstraction or 'convenience' member.
Ideally, this should be removed. Anything called 'ID' should not not carry more context around. Or it needs to be clearly explained in the PathId class description.

Reply:
leaking abstraction often makes implementation easier, but it doesn't mean it is the right thing.

I'll have a look at the CC file, but it currently smells like a convenience hack.